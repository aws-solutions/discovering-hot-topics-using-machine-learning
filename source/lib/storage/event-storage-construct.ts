#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as glue_alpha from '@aws-cdk/aws-glue-alpha';
import { KinesisFirehoseToS3 } from '@aws-solutions-constructs/aws-kinesisfirehose-s3';
import { buildLambdaFunction } from '@aws-solutions-constructs/core';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as firehose from 'aws-cdk-lib/aws-kinesisfirehose';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface EventStorageProps {
    readonly compressionFormat: string;
    /**
     * Should the delivery stream enable dynamic partitioning. If the value is set to true, then either
     * @aggregatebyDay should be set to 'true' or @prefix property should be provided to prefix S3 bucket
     */
    readonly enableDynamicPartitioning?: boolean;
    /**
     * Prefix for S3 bucket, all data will be stored with the provided prefix
     */
    readonly prefix?: string;
    /**
     * If this property is set, it will create a prefix 'created_at=YYYY-MM-DD' for records stored in S3.
     * This property is used in conjunction with the prefix. If the prefix is not set, this property is ignored
     */
    readonly aggregateByDay?: boolean;
    /**
     * Lambda process if any tranformation is required
     */
    readonly processor?: boolean;
    /**
     * S3 bucket location where data will be stored
     */
    readonly s3Bucket: s3.Bucket;
    /**
     * The KMS key ARN for Glue Tables
     */
    readonly keyArn?: string;

    //below props only required if data format is parquet
    /**
     * Convert data to Apache Parquet format
     */
    readonly convertData?: boolean;

    /**
     * The database name is required if Firehose should convert data to Apache Parquet
     */
    readonly database?: glue_alpha.IDatabase;

    /**
     * Table is required to add dependency for Firehose creation
     */
    readonly tableName?: string;
}

export class EventStorage extends Construct {
    private readonly _lambda: lambda.Function;
    private _firehose: firehose.CfnDeliveryStream;

    constructor(scope: Construct, id: string, props: EventStorageProps) {
        super(scope, id);

        // Setup the IAM Role for Kinesis Firehose
        const firehoseRole = new iam.Role(this, 'FirehoseRole', {
            assumedBy: new iam.ServicePrincipal('firehose.amazonaws.com')
        });

        props.s3Bucket.grantReadWrite(firehoseRole); // add permissions to read/write to S3 bucket

        let firehoseGluePolicy: iam.Policy;
        if (props.convertData && props.tableName !== undefined) {
            firehoseGluePolicy = new iam.Policy(this, 'FirehoseGlueTablePolicy', {
                statements: [
                    new iam.PolicyStatement({
                        effect: iam.Effect.ALLOW,
                        actions: ['glue:GetTable', 'glue:GetTableVersion', 'glue:GetTableVersions'],
                        resources: [
                            `arn:aws:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:table/${
                                props.database!.databaseName
                            }/${props.tableName}`,
                            `arn:aws:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:database/${
                                props.database!.databaseName
                            }`,
                            `arn:aws:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:catalog`
                        ]
                    })
                ]
            });

            firehoseGluePolicy.attachToRole(firehoseRole);
        }

        let firehoseGlueKmsPolicy: iam.Policy;
        if (props.keyArn !== undefined) {
            firehoseGlueKmsPolicy = new iam.Policy(this, 'FirehoseGlueKms', {
                statements: [
                    new iam.PolicyStatement({
                        effect: iam.Effect.ALLOW,
                        resources: [props.keyArn],
                        actions: ['kms:Decrypt']
                    })
                ]
            });

            firehoseGlueKmsPolicy.attachToRole(firehoseRole);
        }

        let processingConfiguration = undefined;
        if (props?.processor) {
            if (props.enableDynamicPartitioning) {
                processingConfiguration = {
                    processingConfiguration: {
                        enabled: true,
                        processors: [
                            {
                                type: 'MetadataExtraction',
                                parameters: [
                                    {
                                        parameterName: 'MetadataExtractionQuery',
                                        parameterValue:
                                            '{created_at: .created_at | strptime("%Y-%m-%d %H:%M:%S") | strftime("%Y-%m-%d")}'
                                    },
                                    {
                                        parameterName: 'JsonParsingEngine',
                                        parameterValue: 'JQ-1.6'
                                    }
                                ]
                            }
                        ]
                    }
                };
            } else {
                this._lambda = buildLambdaFunction(this, {
                    lambdaFunctionProps: {
                        runtime: lambda.Runtime.NODEJS_18_X,
                        handler: 'index.handler',
                        code: lambda.Code.fromAsset(`${__dirname}/../../lambda/storage-firehose-processor`)
                    }
                });

                const lambdaProcessorPolicy = new iam.Policy(this, 'LambdaProcessor', {
                    statements: [
                        new iam.PolicyStatement({
                            effect: iam.Effect.ALLOW,
                            resources: [this._lambda.functionArn],
                            actions: ['lambda:InvokeFunction', 'lambda:GetFunctionConfiguration']
                        })
                    ]
                });

                lambdaProcessorPolicy.attachToRole(firehoseRole);
                processingConfiguration = {
                    processingConfiguration: {
                        enabled: true,
                        processors: [
                            {
                                type: 'Lambda',
                                parameters: [
                                    {
                                        parameterName: 'LambdaArn',
                                        parameterValue: this._lambda.functionArn
                                    }
                                ]
                            }
                        ]
                    }
                };
            }
        }

        let bucketPrefix = undefined;
        let dynamicPartitioningConfiguration = undefined;
        if (props.prefix) {
            bucketPrefix = this.getBucketPrefix(props);
            if (props.enableDynamicPartitioning) {
                dynamicPartitioningConfiguration = {
                    dynamicPartitioningConfiguration: {
                        enabled: true,
                        retryOptions: {
                            durationInSeconds: 300
                        }
                    }
                };
            }
        }

        // Setup the default Kinesis Firehose props
        const defaultKinesisFirehoseProps: firehose.CfnDeliveryStreamProps = {
            deliveryStreamType: 'DirectPut',
            deliveryStreamEncryptionConfigurationInput: {
                keyType: 'AWS_OWNED_CMK'
            },
            extendedS3DestinationConfiguration: {
                bucketArn: props.s3Bucket.bucketArn,
                bufferingHints: {
                    intervalInSeconds: 600,
                    sizeInMBs: 128
                },
                compressionFormat: props.compressionFormat,
                roleArn: firehoseRole.roleArn,
                ...dynamicPartitioningConfiguration,
                ...bucketPrefix,
                ...processingConfiguration,
                ...(props.convertData && {
                    dataFormatConversionConfiguration: {
                        inputFormatConfiguration: {
                            deserializer: {
                                openXJsonSerDe: {}
                            }
                        },
                        outputFormatConfiguration: {
                            serializer: {
                                parquetSerDe: {}
                            }
                        },
                        schemaConfiguration: {
                            databaseName: props.database!.databaseName,
                            tableName: props.tableName,
                            roleArn: firehoseRole.roleArn
                        }
                    }
                })
            }
        };

        const firehoseToS3 = new KinesisFirehoseToS3(this, 'KinesisFirehose', {
            kinesisFirehoseProps: defaultKinesisFirehoseProps,
            existingBucketObj: props.s3Bucket
        });

        this._firehose = firehoseToS3.kinesisFirehose;
        this._firehose.node.addDependency(firehoseRole);
        if (props.convertData && props.tableName !== undefined) {
            this._firehose.node.addDependency(firehoseGluePolicy!);
            if (props.keyArn != undefined) {
                this._firehose.node.addDependency(firehoseGlueKmsPolicy!);
            }
        }
    }

    public getBucketPrefix(props: EventStorageProps) {
        let bucketPrefix = undefined;
        if (props.aggregateByDay) {
            bucketPrefix = {
                // updating error prefix and moving it outside the table to avoid the crawler
                // crawling the error prefix
                errorOutputPrefix: `error/${props.prefix}!{firehose:random-string}/!{firehose:error-output-type}/created_at=!{timestamp:yyyy-MM-dd}/`,
                prefix: `${props.prefix}created_at=!{partitionKeyFromQuery:created_at}/`
            };
        } else {
            bucketPrefix = {
                prefix: props.prefix
            };
        }
        return bucketPrefix;
    }

    public get kinesisFirehose(): firehose.CfnDeliveryStream {
        return this._firehose;
    }

    public get deliveryStreamName(): string {
        return this.kinesisFirehose.ref;
    }

    public get deliveryStreamArn(): string {
        return this.kinesisFirehose.attrArn;
    }
}
