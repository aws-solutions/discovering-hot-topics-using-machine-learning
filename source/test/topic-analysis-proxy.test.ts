/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

import { Template } from 'aws-cdk-lib/assertions';
import { Bucket, BucketEncryption } from 'aws-cdk-lib/aws-s3';
import { Stack } from 'aws-cdk-lib';
import { TopicAnalysisProxy } from '../lib/integration/topic-analysis-proxy';
import { EventStorage } from '../lib/storage/event-storage-construct';


test('test Text Analysis Fireshose Stream Creation', () => {
    const stack = new Stack();
    new TopicAnalysisProxy(stack, 'TestTAProxy', {
        topicsStorage: new EventStorage(stack, 'Topic', {
            compressionFormat: 'UNCOMPRESSED',
            prefix: '/topic',
            s3Bucket: new Bucket(stack, 'topics', {
                encryption: BucketEncryption.S3_MANAGED
            })
        }),
        topicMappingsStorage: new EventStorage(stack, 'Mappings', {
            compressionFormat: 'UNCOMPRESSED',
            prefix: '/mappings',
            s3Bucket: new Bucket(stack, 'mappingsbucket', {
                encryption: BucketEncryption.S3_MANAGED
            })
        }),
        topicMappingsInfNS: 'com.analyze.inferece.topic',
        topicsAnalysisInfNS: 'com.analyze.inference.mapping'

    });
    });
