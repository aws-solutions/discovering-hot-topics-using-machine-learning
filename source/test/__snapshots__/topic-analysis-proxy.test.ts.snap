// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Text Analysis Fireshose Stream Creation 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73ArtifactHash7CA57BB1": Object {
      "Description": "Artifact hash for asset \\"358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73\\"",
      "Type": "String",
    },
    "AssetParameters358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73S3Bucket7405AE0E": Object {
      "Description": "S3 bucket for asset \\"358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73\\"",
      "Type": "String",
    },
    "AssetParameters358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73S3VersionKey2BFA9B95": Object {
      "Description": "S3 key for asset version \\"358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MappingsFirehoseRoleC66C52FA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MappingsFirehoseRoleDefaultPolicy997DCB56": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "mappingsbucket4AF7BD5E",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "mappingsbucket4AF7BD5E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MappingsFirehoseRoleDefaultPolicy997DCB56",
        "Roles": Array [
          Object {
            "Ref": "MappingsFirehoseRoleC66C52FA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MappingsKinesisFirehose5E0D6FFA": Object {
      "DependsOn": Array [
        "MappingsFirehoseRoleDefaultPolicy997DCB56",
        "MappingsFirehoseRoleC66C52FA",
      ],
      "Properties": Object {
        "DeliveryStreamEncryptionConfigurationInput": Object {
          "KeyType": "AWS_OWNED_CMK",
        },
        "DeliveryStreamType": "DirectPut",
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "mappingsbucket4AF7BD5E",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 600,
            "SizeInMBs": 128,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "MappingsKinesisFirehosefirehoseloggroupBE9AD2E0",
            },
            "LogStreamName": Object {
              "Ref": "MappingsKinesisFirehosefirehoseloggroupfirehoselogstreamD2508995",
            },
          },
          "CompressionFormat": "UNCOMPRESSED",
          "EncryptionConfiguration": Object {
            "KMSEncryptionConfig": Object {
              "AWSKMSKeyARN": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":kms:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":alias/aws/s3",
                  ],
                ],
              },
            },
          },
          "Prefix": "/mappings",
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "MappingsFirehoseRoleC66C52FA",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "MappingsKinesisFirehoseKinesisFirehosePolicy68E53706": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:AbortMultipartUpload",
                "s3:GetBucketLocation",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "mappingsbucket4AF7BD5E",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "mappingsbucket4AF7BD5E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "MappingsKinesisFirehosefirehoseloggroupBE9AD2E0",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "MappingsKinesisFirehosefirehoseloggroupfirehoselogstreamD2508995",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MappingsKinesisFirehoseKinesisFirehosePolicy68E53706",
        "Roles": Array [
          Object {
            "Ref": "MappingsKinesisFirehoseKinesisFirehoseRole4C9908D7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MappingsKinesisFirehoseKinesisFirehoseRole4C9908D7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MappingsKinesisFirehosefirehoseloggroupBE9AD2E0": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W86",
              "reason": "Retention period for CloudWatchLogs LogGroups are set to 'Never Expire' to preserve customer data indefinitely",
            },
            Object {
              "id": "W84",
              "reason": "By default CloudWatchLogs LogGroups data is encrypted using the CloudWatch server-side encryption keys (AWS Managed Keys)",
            },
          ],
        },
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "MappingsKinesisFirehosefirehoseloggroupfirehoselogstreamD2508995": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "MappingsKinesisFirehosefirehoseloggroupBE9AD2E0",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "TestTAProxyLambdaFunction646B9726": Object {
      "DependsOn": Array [
        "TestTAProxyLambdaFunctionServiceRoleDefaultPolicyFC7C1A09",
        "TestTAProxyLambdaFunctionServiceRole1AE0D1E2",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with tighter permissions.",
            },
            Object {
              "id": "W89",
              "reason": "This is not a rule for the general case, just for specific use cases/industries",
            },
            Object {
              "id": "W92",
              "reason": "Impossible for us to define the correct concurrency for clients",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73S3Bucket7405AE0E",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73S3VersionKey2BFA9B95",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters358be7f4c2e7dbf942b8403404cafa903a2068b7333c4585fb619999b41eab73S3VersionKey2BFA9B95",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "TOPICS_FIREHOSE": Object {
              "Ref": "TopicKinesisFirehose8F663708",
            },
            "TOPICS_NS": "com.analyze.inference.mapping",
            "TOPIC_MAPPINGS_FIREHOSE": Object {
              "Ref": "MappingsKinesisFirehose5E0D6FFA",
            },
            "TOPIC_MAPPINGS_NS": "com.analyze.inferece.topic",
          },
        },
        "Handler": "lambda_function.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestTAProxyLambdaFunctionServiceRole1AE0D1E2",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 600,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestTAProxyLambdaFunctionServiceRole1AE0D1E2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestTAProxyLambdaFunctionServiceRoleDefaultPolicyFC7C1A09": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray and access ENIs in a VPC.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "firehose:PutRecord",
                "firehose:PutRecordBatch",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":firehose:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":deliverystream/",
                      Object {
                        "Ref": "TopicKinesisFirehose8F663708",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":firehose:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":deliverystream/",
                      Object {
                        "Ref": "MappingsKinesisFirehose5E0D6FFA",
                      },
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestTAProxyLambdaFunctionServiceRoleDefaultPolicyFC7C1A09",
        "Roles": Array [
          Object {
            "Ref": "TestTAProxyLambdaFunctionServiceRole1AE0D1E2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TopicFirehoseRoleCE52DEAF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TopicFirehoseRoleDefaultPolicyD8DD6E87": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "topics8661C816",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "topics8661C816",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TopicFirehoseRoleDefaultPolicyD8DD6E87",
        "Roles": Array [
          Object {
            "Ref": "TopicFirehoseRoleCE52DEAF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TopicKinesisFirehose8F663708": Object {
      "DependsOn": Array [
        "TopicFirehoseRoleDefaultPolicyD8DD6E87",
        "TopicFirehoseRoleCE52DEAF",
      ],
      "Properties": Object {
        "DeliveryStreamEncryptionConfigurationInput": Object {
          "KeyType": "AWS_OWNED_CMK",
        },
        "DeliveryStreamType": "DirectPut",
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "topics8661C816",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 600,
            "SizeInMBs": 128,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "TopicKinesisFirehosefirehoseloggroupAA816077",
            },
            "LogStreamName": Object {
              "Ref": "TopicKinesisFirehosefirehoseloggroupfirehoselogstreamB6B49275",
            },
          },
          "CompressionFormat": "UNCOMPRESSED",
          "EncryptionConfiguration": Object {
            "KMSEncryptionConfig": Object {
              "AWSKMSKeyARN": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":kms:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":alias/aws/s3",
                  ],
                ],
              },
            },
          },
          "Prefix": "/topic",
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "TopicFirehoseRoleCE52DEAF",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "TopicKinesisFirehoseKinesisFirehosePolicyEC75E12A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:AbortMultipartUpload",
                "s3:GetBucketLocation",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "topics8661C816",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "topics8661C816",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "TopicKinesisFirehosefirehoseloggroupAA816077",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "TopicKinesisFirehosefirehoseloggroupfirehoselogstreamB6B49275",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TopicKinesisFirehoseKinesisFirehosePolicyEC75E12A",
        "Roles": Array [
          Object {
            "Ref": "TopicKinesisFirehoseKinesisFirehoseRoleA113EEB3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TopicKinesisFirehoseKinesisFirehoseRoleA113EEB3": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TopicKinesisFirehosefirehoseloggroupAA816077": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W86",
              "reason": "Retention period for CloudWatchLogs LogGroups are set to 'Never Expire' to preserve customer data indefinitely",
            },
            Object {
              "id": "W84",
              "reason": "By default CloudWatchLogs LogGroups data is encrypted using the CloudWatch server-side encryption keys (AWS Managed Keys)",
            },
          ],
        },
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TopicKinesisFirehosefirehoseloggroupfirehoselogstreamB6B49275": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "TopicKinesisFirehosefirehoseloggroupAA816077",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "mappingsbucket4AF7BD5E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "topics8661C816": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
